### PR TITLE
implementation of spin1_dma_flush

### PR DESCRIPTION
**Pull request by lplana**
This pull request implements function 'spi1_dma_flush', requested in Issue #23.

I believe that this implementation completes all the tasks listed in the issue.

### CODE REVIEW

File: spin1_api/spin1_api.c [hunk 1]
```diff
@@ -1087,6 +1087,75 @@ uint spin1_dma_transfer(uint tag, void *system_address, void *tcm_address,
...
+    // flush the software DMA transfer queue,
+    uint cpsr = spin1_int_disable();
+    dma_queue.start = 0;
+    dma_queue.end   = 0;
+    spin1_mode_restore(cpsr);
+
+    // abort any ongoing transfer in the DMA controller,
+    // flush the hardware queue in the DMA controller,
+    // and clear any pending DMA_COMPLETE interrupts in the DMA controller
+    dma[DMA_CTRL] = 0x1f;
```
Comment (rowleya) [inline spin1_api/spin1_api.c]: The above clears the DMA queue with interrupts disabled.  What happens if an interrupt happens just before we clear?  Would this cause issues?

File: spin1_api/spin1_api.c [hunk 2]
```diff
@@ -1087,6 +1087,75 @@ uint spin1_dma_transfer(uint tag, void *system_address, void *tcm_address,
...
+                        mtp = (mtp + 1) % TASK_QUEUE_SIZE;
+                    }
+                    mfp = (mfp + 1) % TASK_QUEUE_SIZE;
+                }
+
+                // update queue end pointer
+                tq->end = mtp;
+            }
+        }
+        spin1_mode_restore(cpsr);
```
Comment (rowleya) [inline spin1_api/spin1_api.c:1076]: Could an interrupt not get in here and add a new item to the queue?

File: spin1_api/spin1_api.c (hunk 1 above)
Comment (lplana) [reply spin1_api/spin1_api.c]: Both of your comments are correct. Concurrent threads started by interrupts can trigger or request new DMA transfers. For that reason, the function attempts to abort/flush 'as much as possible' while making sure that all queues are kept in a consistent state. I believe that the integrity of the queues is not compromised by the arrival of interrupts. It's up to the user to deal with requests that 'creep in'.

Looking at the function again I believe that clearing/aborting the DMA controller at the top of the function should reduce the possibility of interrupts arriving. If really pushing it, the whole function can be run with interrupts disabled to completely avoid them.

File: spin1_api/spin1_api.c (hunk 2 above)
Comment (lplana) [reply spin1_api/spin1_api.c:1076]: See comment above.

**Review by rowleya — APPROVED** [review]

**Comment by rowleya** [comment]
I have approved this, but then noticed that the suggestion to move the DMA reset to the top of the function hadn't been done.  Is this still to be done, or has it been decided that it shouldn't?

**Pull request merged**